Fix `read_exact` and `read_buf_exact` for `&[u8]` and `io:Cursor`

a1phyr opened the pull request:
- Drain after `read_exact` and `read_buf_exact` 
- Append to cursor in `read_buf_exact`

a1phyr committed:
Fix `read_exact` and `read_buf_exact` for `&[u8]` and `io:Cursor`

workingjubilee commented on library/std/src/io/cursor.rs:
```diff
@@ -352,17 +352,24 @@ where
+
+        match result {
+            Ok(_) => self.pos += buf.len() as u64,
+            // The only posible error condition is EOF
```
Typo, also we might as well finish the thought for the reader:
```suggestion
            // The only possible error condition is EOF, so place the cursor at "EOF"
```

workingjubilee commented on library/std/src/io/cursor.rs:
```diff
@@ -352,17 +352,24 @@ where
-        Read::read_exact(&mut self.remaining_slice(), buf)?;
-        self.pos += n as u64;
-        Ok(())
+        let result = Read::read_exact(&mut self.remaining_slice(), buf);
```
Can you add a doctest to the implementation here that demonstrates the behavior?

workingjubilee commented on library/std/src/io/cursor.rs:
```diff
@@ -352,17 +352,24 @@ where
-        Read::read_buf_exact(&mut self.remaining_slice(), cursor)?;
-        self.pos += n as u64;
-        Ok(())
+    fn read_buf_exact(&mut self, mut cursor: BorrowedCursor<'_>) -> io::Result<()> {
```
Doctest?

workingjubilee commented on library/std/src/io/impls.rs:
```diff
@@ -307,6 +310,9 @@ impl Read for &[u8] {
     #[inline]
     fn read_buf_exact(&mut self, mut cursor: BorrowedCursor<'_>) -> io::Result<()> {
         if cursor.capacity() > self.len() {
+            // Append everything we can to the cursor.
+            cursor.append(*self);
+            *self = &self[self.len()..];
```
is there a reasonable doctest to demonstrate this behavior? hmmm...

workingjubilee commented on library/std/src/io/impls.rs:
```diff
@@ -287,6 +287,9 @@ impl Read for &[u8] {
     #[inline]
     fn read_exact(&mut self, buf: &mut [u8]) -> io::Result<()> {
         if buf.len() > self.len() {
+            // `read_exact` makes no promise about the content of `buf` if it
+            // fails so don't bother about that.
+            *self = &self[self.len()..];
```
doct... actually I'm pretty sure there isn't a useful doctest for "lol no promises", right? don't worry about this one, if so.

a1phyr committed:
Address review comments

a1phyr commented:
I've added tests for the exact behavior of the readers on EOF.

I don't think that doctests are useful here so I've added normal unit tests instead. Please tell me if you think that we doctests are what we really need.

workingjubilee commented:
hmm, that's fine by me!

@bors r+

bors commented:
:pushpin: Commit a197ff325989ed200fc4730d35501b9e4064316b has been approved by `workingjubilee`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=workingjubilee a197ff325989ed200fc4730d35501b9e4064316b -->
<!-- homu: {"type":"Approved","sha":"a197ff325989ed200fc4730d35501b9e4064316b","approver":"workingjubilee","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
